kinesis/kinesis.go

```diff
@@ -637,6 +646,79 @@ func gzipCompress(data []byte) ([]byte, error) {
 	return b.Bytes(), nil
 }
 
+func compressThenTruncate(compressorFunc CompressorFunc, data []byte, maxOutLen int, truncatedSuffix []byte, outputPlugin OutputPlugin) ([]byte, error) {
+
+	truncationReductionPercent := 90
+	var compressedData []byte
+	var truncationBuffer []byte
+	var originalCompressedLen int
+	var compressedLen int
+	var err error
+
+	/* Iterative approach to truncation */
+	isTruncated := false
+	compressedLen = math.MaxInt64
+	truncatedInLen := len(data)
+	truncationBuffer = data
+	for (maxOutLen < compressedLen) {
+		compressedData, err = compressorFunc(truncationBuffer)
+		if err != nil {
+			return nil, err
+		}
+		compressedLen = len(compressedData)
+
+		/* Truncation needed */
+		if (maxOutLen < compressedLen) {
+			logrus.Debugf("[kinesis %d] iterative truncation round stream=%s\n",
+						 outputPlugin.PluginID, outputPlugin.stream)
+
+			/* Base case: input compressed empty string, output still too large */
+			if (truncatedInLen == 0) {
+				logrus.Errorf("[kinesis %d] truncation failed, compressed empty input too " +
+							 "large stream=%s\n", outputPlugin.PluginID, outputPlugin.stream)
+				return nil, errors.New("compressed empty to large");
+			}
+
+			/* Calculate corrected input size */
+			truncatedInLenPrev := truncatedInLen;
+			truncatedInLen = (maxOutLen * truncatedInLen) / compressedLen;
+			truncatedInLen = (truncatedInLen * truncationReductionPercent) / 100;
+
+			/* Ensure working down */
+			if (truncatedInLen >= truncatedInLenPrev) {
+				truncatedInLen = truncatedInLenPrev - 1;
+			}
+
+			/* Allocate truncation buffer */
+			if (!isTruncated) {
+				isTruncated = true;
+				originalCompressedLen = compressedLen
+				truncationBuffer = make([]byte, truncatedInLen)
+				copy(truncationBuffer, data[:truncatedInLen])
+			}
+
+			/* Slap on truncation suffix */
+			if (truncatedInLen < len(truncatedSuffix)) {
+				/* No room for the truncation suffix. Terminal error */
+				logrus.Errorf("[kinesis %d] truncation failed, no room for suffix " +
+							 "stream=%s\n", outputPlugin.PluginID, outputPlugin.stream)
+				return nil, errors.New("no room for suffix");
+			}
+			truncationBuffer = truncationBuffer[:truncatedInLen]
+			copy(truncationBuffer[len(truncationBuffer)-len(truncatedSuffix):], truncatedSuffix)
+		}
+	}
+
+	/* Truncate buffer free and compression buffer allocation */
```

Comment:
I didn't get what this comment mean. The code following this comment is just a log warn. No related actions match the comment.

Reply:
Shucks, I copied the c plugin code, and updated line for line. I guess since go has a garbage collector, and the flb_frees were removed, this comment needs to be updated.